builder/proxmox/common/step_start_vm.go

```diff
@@ -277,33 +279,94 @@ func generateProxmoxNetworkAdapters(nics []NICConfig) proxmox.QemuDevices {
 	}
 	return devs
 }
-func generateProxmoxDisks(disks []diskConfig) proxmox.QemuDevices {
-	devs := make(proxmox.QemuDevices)
-	for idx := range disks {
-		devs[idx] = make(proxmox.QemuDevice)
-		setDeviceParamIfDefined(devs[idx], "type", disks[idx].Type)
-		setDeviceParamIfDefined(devs[idx], "size", disks[idx].Size)
-		setDeviceParamIfDefined(devs[idx], "storage", disks[idx].StoragePool)
-		setDeviceParamIfDefined(devs[idx], "cache", disks[idx].CacheMode)
-		setDeviceParamIfDefined(devs[idx], "format", disks[idx].DiskFormat)
-
-		if (devs[idx]["type"] == "scsi" || devs[idx]["type"] == "virtio") &&
-			disks[idx].IOThread {
-			devs[idx]["iothread"] = "true"
-		}
 
-		if disks[idx].Discard {
-			devs[idx]["discard"] = "on"
-		} else {
-			devs[idx]["discard"] = "ignore"
+func generateProxmoxDisks(disks []diskConfig) *proxmox.QemuStorages {
+	ideDisks := proxmox.QemuIdeDisks{}
+	sataDisks := proxmox.QemuSataDisks{}
+	scsiDisks := proxmox.QemuScsiDisks{}
+	virtIODisks := proxmox.QemuVirtIODisks{}
+
+	ideCount := 0
+	sataCount := 0
+	scsiCount := 0
+	virtIOCount := 0
+
+	for idx := range disks {
+		tmpSize, _ := strconv.ParseInt(disks[idx].Size[:len(disks[idx].Size)-1], 10, 0)
+		size := proxmox.QemuDiskSize(0)
+		switch disks[idx].Size[len(disks[idx].Size)-1:] {
+		case "T":
+			size = proxmox.QemuDiskSize(tmpSize) * 1073741824
+		case "G":
+			size = proxmox.QemuDiskSize(tmpSize) * 1048576
+		case "M":
+			size = proxmox.QemuDiskSize(tmpSize) * 1024
+		case "K":
+			size = proxmox.QemuDiskSize(tmpSize)
 		}
 
-		// Add SSD flag only if true
-		if disks[idx].SSD {
-			devs[idx]["ssd"] = "1"
+		switch disks[idx].Type {
+		case "ide":
+			dev := proxmox.QemuIdeStorage{
+				Disk: &proxmox.QemuIdeDisk{
+					SizeInKibibytes: size,
+					Storage:         disks[idx].StoragePool,
+					Cache:           proxmox.QemuDiskCache(disks[idx].CacheMode),
+					Format:          proxmox.QemuDiskFormat(disks[idx].DiskFormat),
+					Discard:         disks[idx].Discard,
+					EmulateSSD:      disks[idx].SSD,
+				},
+			}
+			reflect.ValueOf(&ideDisks).Elem().FieldByIndex([]int{ideCount}).Set(reflect.ValueOf(&dev))
```

Comment:
I'm not sure I understand why reflection is needed here, I was focused on the counts at my first review so much that I missed this but this looks weird :sweat_smile: 
If there's a good reason for this, I would suggest documenting why, as it's not immediately clear I think.

Reply:
I've added comments to each one to explain its purpose. 
The Telmate/proxmox-api-go/proxmox doesn't provide any functions to assign devices to the parent struct fields so I've used reflect to walk the parent struct and assign the generated disk to a field matching the count integer value, eg. the first ide disk defined gets assigned to &ideDisks.Disk_0, second ide disk to &ideDisks.Disk_1 and so on again to align with the behaviour of the previous generateProxmoxDisks code behaviour.

There is also validation in config.go to make sure you can't define more disks than the parent struct can handle, eg &ideDisks is a proxmox.QemuIdeDisks{} struct, which only has 4 Disk_<_x_> fields defined, two of which are used for ISOs

Reply:
Hey @lbajolet-hashicorp!

Not sure why I didn't think of using the field name for the reflection 😅 it reads better and makes sure the dev is assigned to a Disk_x field in the event new fields for a different purpose are added to the Qemu<device type>Disks structs. I also appreciate the extra tests and commenting of the reflection, looks much better.

Thanks for the commits, LGTM!

Reply:
Thanks for the update @mpywell, glad you think the commits I added make sense!

Will merge this right now